### PR TITLE
BREAKING CHANGE: Big rework

### LINKED ISSUE #2: Add documentation
Improve documentation to better describe how to use the MMM and what is can be used for.

### PR DESCRIPTION
**Pull request by MrKevinWeiss**
This is pretty much a rewrite of the whole system.
There are some things slightly preserved such as the csv and general C
output.
However, the source code is rewritten largely due to the knowledge of the
author growing (or changing).

The rewrite simplifies some aspect of the system due to the following:
- Generated code uses jinja templates
- Classes are used that allow members to hold state instead of passing
args everywhere
- Recursion is used much less

Though complexity of the previous implementation might have been
reduced, even more complexity is in form of additional features:
- **yaml based map config vs json** -> It is much easier to write and read yaml
than json as well as the ability comment makes it worth it.
- **Multiple maps files can be combined** -> This can simplify the more complex
as one large file can be hard to read. Also, this helps to deduplicate maps
as some map configurations can be shared.
- **Map configuration output generated in different files** -> Previously the
output of the map altered the actual config file, allowing overrides but making
it difficult for source control and managing overrides.
- **Overrides in map config file** -> Since the single file containing all the
input and output data is not longer used, the map configuration file can
specify overrides there. This is much more explicit and readable than before.
- **Enumerations** -> Support is added for enumerations now, though they
are not very functional in the generated C, python tools will benefit down the
road.
- **Access/permissions** -> The access rework now integrates friendlier names
in the form of permission users.
- **Dependency resolution** -> Now, position doesn't matter, as long as
something can get resolved, it will. The generated files are also split
and include the dependencies.
- **Unique IDs** -> Previously the version was somewhat used to identify the
may. However, this is a problem, specifically when there are multiple devices
that use this and the interface gets confused which memory map should be used.
Long story short, we now go off hashes which are a better way to identify
maps.
- **Tests** -> Many many more tests have been added, so much so that we have
100% coverage.
- **Examples** -> Also some better examples, even some that compile C have been
added.
- **Defines** -> Defines that can be resolved not only help readability but
allow variants to be easily declared.

Fixes #2 #11 #19 #22 #27 #28 #29

### CODE REVIEW

**Comment by MrKevinWeiss** [comment]
Got to start somewhere... Again...
